pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -1133,14 +1135,31 @@ func (m *kubeGenericRuntimeManager) computeInitContainerActions(pod *v1.Pod, pod
 					if liveness == proberesults.Failure {
 						// If the restartable init container failed the liveness probe,
 						// restart it.
+						message = fmt.Sprintf("Init container %s failed liveness probe, will be restarted", container.Name)
 						changes.ContainersToKill[status.ID] = containerToKillInfo{
 							name:      container.Name,
 							container: container,
-							message:   fmt.Sprintf("Init container %s failed liveness probe", container.Name),
+							message:   message,
 							reason:    reasonLivenessProbe,
 						}
 						changes.InitContainersToStart = append(changes.InitContainersToStart, i)
+						break
+					}
+				}
+
+				// Restart running sidecar containers which have had their definition changed.
```

Comment:
one interesting scenario: imagine that the container is in startup probe and cannot pass the startup probe because of a bug. Operator updates the image. Should kubelet reload the container and restart the startup probe? Or only reload it after startup probe has succeeded?

My quick look at this code suggests that startup probe execution will block this reload because of the image. Is it consistent with how regular containers behave?

Reply:
From my understanding, if the probe returns failed the init container will be restarted and the new image will be pulled. The refactored initcontainerchanges method attempts to make it clear that no matter what reason the init container fails, it will be restarted. Adding a test case for this is a good idea however to make sure this scenario works for both types of containers, sidecar and regular

Reply:
As a side note, if the startup probe or liveness probe fails, should we call the PostStop hook before restarting the container? That is the [current behavior for regular containers](https://github.com/kubernetes/kubernetes/blob/dd4943c831dbc43fb8d410ccdbc27e3c807131fa/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L960)

Reply:
> if the probe returns failed the init container will be restarted and the new image will be pulled.

Failed startup probe mean we need to keep waiting for it to initialize. Container will not be restarted because of a failed startup probe. Thus the question.

If you can cover this scenario with the test - it will be great